File: src/treemapper/diffctx/edges/structural/containment.py
from __future__ import annotations

from collections import defaultdict
from pathlib import Path

from ...config import EDGE_WEIGHTS
from ...types import Fragment, FragmentId
from ..base import EdgeBuilder, EdgeDict


class ContainmentEdgeBuilder(EdgeBuilder):
    weight = EDGE_WEIGHTS["containment"].forward
    reverse_weight_factor = EDGE_WEIGHTS["containment"].reverse_factor

    def build(self, fragments: list[Fragment], repo_root: Path | None = None) -> EdgeDict:
        by_path: dict[Path, list[Fragment]] = defaultdict(list)
        for f in fragments:
            by_path[f.path].append(f)

        edges: EdgeDict = {}

        for _path, frags in by_path.items():
            if len(frags) >= 2:
                self._process_file_containment(frags, edges)

        return edges

    def _process_file_containment(self, frags: list[Fragment], edges: EdgeDict) -> None:
        frags_sorted = sorted(frags, key=lambda x: (x.start_line, -x.end_line))
        stack: list[Fragment] = []

        for f in frags_sorted:
            while stack and f.start_line > stack[-1].end_line:
                stack.pop()

            if stack:
                self._add_containment_edge(f, stack[-1], edges)

            stack.append(f)

    def _add_containment_edge(self, child: Fragment, parent: Fragment, edges: EdgeDict) -> None:
        if parent.start_line <= child.start_line and child.end_line <= parent.end_line and parent.id != child.id:
            self.add_edge(edges, child.id, parent.id)


def _build_containment_edges(fragments: list[Fragment]) -> dict[tuple[FragmentId, FragmentId], float]:
    return ContainmentEdgeBuilder().build(fragments)

File: diffctx/src/edges/structural/containment.rs
use std::path::Path;

use rustc_hash::FxHashMap;

use crate::config::weights::EDGE_WEIGHTS;
use crate::types::Fragment;

use super::super::EdgeDict;
use super::super::base::{EdgeBuilder, add_edge};

pub struct ContainmentEdgeBuilder;

impl EdgeBuilder for ContainmentEdgeBuilder {
    fn build(&self, fragments: &[Fragment], _repo_root: Option<&Path>) -> EdgeDict {
        let weight = EDGE_WEIGHTS["containment"].forward;
        let reverse_factor = EDGE_WEIGHTS["containment"].reverse_factor;

        let mut by_path: FxHashMap<&str, Vec<&Fragment>> = FxHashMap::default();
        for f in fragments {
            by_path.entry(f.path()).or_default().push(f);
        }

        let mut edges: EdgeDict = FxHashMap::default();

        for (_path, frags) in &by_path {
            if frags.len() < 2 {
                continue;
            }
            let mut sorted = frags.clone();
            sorted.sort_by(|a, b| {
                a.start_line()
                    .cmp(&b.start_line())
                    .then(b.end_line().cmp(&a.end_line()))
            });

            let mut stack: Vec<&Fragment> = Vec::new();

            for f in &sorted {
                while let Some(top) = stack.last() {
                    if f.start_line() > top.end_line() {
                        stack.pop();
                    } else {
                        break;
                    }
                }

                if let Some(parent) = stack.last() {
                    if parent.start_line() <= f.start_line()
                        && f.end_line() <= parent.end_line()
                        && parent.id != f.id
                    {
                        add_edge(&mut edges, &f.id, &parent.id, weight, reverse_factor);
                    }
                }

                stack.push(f);
            }
        }

        edges
    }
}
